src/pysdmx/model/vtl.py

```
@@ -25,27 +33,78 @@ class Transformation(Item, frozen=True, omit_defaults=True):
     is_persistent: bool = False
     result: str = ""
 
+    def __post_init__(self) -> None:
+        """Additional validation checks for transformations."""
+        try:
+            create_ast(self.full_expression)
+        except Exception as e:
+            raise Invalid(
+                f"Invalid transformation definition: {str(e)}"
+            ) from e
+
     @property
     def full_expression(self) -> str:
         """Return the full expression with the semicolon."""
         assign_operand = "<-" if self.is_persistent else ":="
-
-        return f"{self.result} {assign_operand} {self.expression};"
+        full_expression = f"{self.result} {assign_operand} {self.expression}"
+        return (
+            full_expression
+            if full_expression.strip().endswith(";")
+            else f"{full_expression};"
+        )
 
 
 class Ruleset(Item, frozen=True, omit_defaults=True):
     """A persistent set of rules."""
 
     ruleset_definition: str = ""
-    ruleset_scope: str = ""
-    ruleset_type: str = ""
+    ruleset_scope: Optional[Literal["variable", "valuedomain"]] = None
+    ruleset_type: Optional[Literal["datapoint", "hierarchical"]] = None
+
+    def __post_init__(self) -> None:
+        """Additional validation checks for rulesets."""
+        try:
+            ast = create_ast(self.ruleset_definition)
+        except Exception as e:
+            raise Invalid(f"Invalid ruleset definition: {str(e)}") from e
+        if self.ruleset_type == "hierarchical" and not isinstance(
+            ast.children[0], ASTHRuleset
+        ):
+            raise Invalid("Ruleset type does not match the definition")
+        if self.ruleset_type == "datapoint" and not isinstance(
+            ast.children[0], ASTDPRuleset
+        ):
+            raise Invalid("Ruleset type does not match the definition")
```

We need to ensure as well that the list of ast children has only 1 item. If there is more than one we should raise an error that a single RulesetDefinition is valid in a Ruleset

---

Added a validation with `len(ast.childen)` to check this